### PR TITLE
boards: arm: xilinx: Add initial KV260 R5 board support

### PR DESCRIPTION
**Pull request by uLipe**
Enables usage of the Arm Cortex-R5 real-time processing unit (RPU) on
the Kria KV260 development board. Only the first R5 core is currently
supported.

### CODE REVIEW

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Why is there a need for a board-specific "Hello, World" sample?

**Comment by microbuilder** [comment]
> Why is there a need for a board-specific "Hello, World" sample?

There doesn't seem to be anything special about the sample as it is today. Other than documentation, which can go in the boards folder, is there a reason the standard hello world sample can't be used instead?

File: samples/boards/kv260_starter_kit_rpu/vitis/settings64.sh
Comment (microbuilder) [inline samples/boards/kv260_starter_kit_rpu/vitis/settings64.sh:17]: This should be updated to avoid hard-coded paths based on `/tools/Xilinx`

File: samples/boards/kv260_starter_kit_rpu/vitis/debug_zephyr.tcl
Comment (microbuilder) [inline samples/boards/kv260_starter_kit_rpu/vitis/debug_zephyr.tcl:6]: Avoid hard-coded `/tools/Xilinx` path

File: samples/boards/kv260_starter_kit_rpu/vitis/debug_zephyr.tcl
Comment (microbuilder) [inline samples/boards/kv260_starter_kit_rpu/vitis/debug_zephyr.tcl:19]: hard-coded path again

File: boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst:72]: Too much white space

**Review by microbuilder — CHANGES_REQUESTED** [review]

**Comment by uLipe** [comment]
Thanks for the comments @henrikbrixandersen @microbuilder , the problem here is to run a zephyr application to the R5 units on the KV260 is the strong dependency from Xilinx tools, that is it the scripts TCL present in this PR.

One thing I can do here, is to just remove the source files and point the cmake to import them from the generic hello world sample, does it sound reasonable?

File: samples/boards/kv260_starter_kit_rpu/vitis/debug_zephyr.tcl
Comment (uLipe) [reply samples/boards/kv260_starter_kit_rpu/vitis/debug_zephyr.tcl:6]: the tools/Xilinx path is always selected from Vitis during installation, of course the 2022.1 part from its path should be handled to avoid problems between specific versions.

**Comment by uLipe** [comment]
This sample is more to demonstrate on how to debug a Zephyr application on the R5 units, the challenge here, is  Xilinx obligates the user to flash and run debug sessions through its own GDB which is behind of Vitis tool.

**Comment by uLipe** [comment]
I cleaned up the PR and removed the debugging sample, the support of debug via vits will be added in another PR.

Please PTAL again.

**Comment by wmamills** [comment]
@uLipe This looks much better w/o the debug support.

The board name is pretty long and there is no value in keeping starter_kit as part of the name.  The only kv260 one can order is the starter kit and everything you have done here applies equally as well to the kv260 with the production SOM (bought separately) as well.

Which UART are we using?  It says UART0 in the readme but the dts and config look like we are using the second one.

The readme should describe how to run it.  The suggestion now is to use the Linux kernel's remote proc to load & run.
Any petalinux (the Xilinx Linux distro) kernel should work as will any upstream kernel v6.1 or later.  Look at the scripts in the OpenAMP demo docker image for the details of the howto.

More comments inline

File: boards/arm/xilinx_kv260_starter_kit_r5/board.cmake
Comment (wmamills) [inline boards/arm/xilinx_kv260_starter_kit_r5/board.cmake:3]: fix newline

File: boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst
Comment (wmamills) [inline boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst:14]: The R5 subsystem on Zynqmp is called the RPU, real-time processing unit.

It can operate as:
* two independent R5 cores with their own TCMs (tightly coupled memories)
* or as a single dual lock step unit with double the TCM size.

====
RPU is a term used in the code so it is good to define it.  The multiple modes of the RPU are pretty fundamental to understanding this platform so give an overview.

The above text is a suggestion.  Please feel free to reword an place at an appropriate spot.

File: boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst
Comment (wmamills) [inline boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst:53]: This is a limitation of the emulator.
Is this really a limitation of the zephyr port?
Can we build lock step and split mode today?  If so how?
If not we can defer to another PR but document which you are doing.

File: boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst
Comment (wmamills) [inline boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst:61]: Where is this running?  I think the below only builds it correct?

**Review by wmamills — CHANGES_REQUESTED** [review]
As I said in the comment, looks much better w/o the debug stuff.

Inline change requests cover most of the points.
However, I dont seem to be able to comment on your commit header.  It should be better.  At least mention AMD Xilinx and the R5 in the short description.

In the body explain that this is initial support and enough to get basic UART samples working and loaded from Linux.

Thanks Felipe, almost there,
Bill

File: boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst
Comment (uLipe) [reply boards/arm/xilinx_kv260_starter_kit_r5/doc/index.rst:53]: Just reworded, it seems a limitation from zephyr port here.

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:65]: Can you add a link to somewhere users can find instructions on downloading/building and flashing the SD card (etc.) with an appropriate Linux image, to at least point them in the right direction?

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:78]:
```suggestion
After a succesful build, deploy the .elf file generated to the
target board. On the target board terminal load the zephyr file
to the desired RPU:
```

**Review by microbuilder — CHANGES_REQUESTED** [review]

**Review by wmamills — CHANGES_REQUESTED** [review]
The documentation is really the only thing holding this back.
See the comments inline.

Also, please write a better commit message.  I said that last time as well.  You can duplicate a bit of the board description here.  Put Cortex-R5 in the title.  You can keep RPU as well.

**Review by wmamills — CHANGES_REQUESTED** [review]
The documentation is really the only thing holding this back.
See the comments inline.

Also, please write a better commit message.  I said that last time as well.  You can duplicate a bit of the board description here.  Put Cortex-R5 in the title.  You can keep RPU as well.

**Review by microbuilder — CHANGES_REQUESTED** [review]

**Comment by microbuilder** [comment]
nit, but please adjust the commit message to start with a capital letter as well. Thanks.

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (uLipe) [reply boards/arm/xilinx_kv260_r5/doc/index.rst]: Hmm this PR is open for a while, in that time it was tested under v2.2

**Comment by microbuilder** [comment]
> nit, but please adjust the commit message to start with a capital letter as well. Thanks.

Please also fix the commit message, as requested. Thanks.

**Comment by uLipe** [comment]
Ahh i just applied the capital letter on the commit header message. Please check if it is now correct.

Thanks.

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (wmamills) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:10]: Add link to a good page for kv260 info in here somewhere.
Match what is done for other boards.

**Review by wmamills — CHANGES_REQUESTED** [review]
The title of the PR is good.  The title of the commit is not.  Make it match the PR title.  You can shorten it if needed.

**Comment by microbuilder** [comment]
> The title of the PR is good. The title of the commit is not. Make it match the PR title. You can shorten it if needed.

I'd suggest both be changed to:

```
boards: arm: xilinx: Add initial KV260 board support
```

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (wmamills) [reply boards/arm/xilinx_kv260_r5/doc/index.rst:10]: As @microbuilder pointed out on our call.  The page should also have a picture of the board to match what is normally done for other boards.

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (uLipe) [reply boards/arm/xilinx_kv260_r5/doc/index.rst:51]: UART1 currently is the instance of the peripheral that appears in the DTS file, in order this is the second one, the first one would be the UART0

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (uLipe) [reply boards/arm/xilinx_kv260_r5/doc/index.rst:52]: Hmm well actually the application is linked and is loaded to the sram0 area, when the DTS is parsed a DT_SRAM_ symbol is generated and consumed by the linker, and if the board does not support XIP execution, all the application code plus data is placed at sram0 which seems to be an DDR area.  

The OCM area is defined on the linker as well, and consumes the ocm0 area from DTS file, but by default nothing is placed there unless the code explicitly tells that (placement via attributes for example).

I will put a section explaining that .

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:103]: Shouldn't this be `zephyr.elf`?

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:99]:
```suggestion
After a successful build, deploy the generated .elf file to the target board.
Using the Linux shell on the target board (available using one of the UART
blocks exposed via the on-board FTDI chip), run the zephyr firmware on
the desired RPU via remoteproc with:
```

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:107]:
```suggestion
With another terminal connected to UART1 on the host machine
(available via one of the tty ports with the on-board FTDI chip),
you should see the Zephyr application running:
```

**Review by microbuilder — CHANGES_REQUESTED** [review]

**Comment by uLipe** [comment]
CI errors don't seem to be related to this PR contents.

**Review by microbuilder — DISMISSED** [review]

File: boards/arm/xilinx_kv260_r5/doc/index.rst
Comment (microbuilder) [inline boards/arm/xilinx_kv260_r5/doc/index.rst:102]: I already approved, so more of a nit for clarity, but this might be easier to follow:

```suggestion
Due to a hardware limitation, both Linux and Zephyr share the same UART
controller, meaning when the Zephyr application is started it will takeover the
console from Linux.

To avoid this limitation when accessing the Linux shell, the best approach is to
connect to the board using ``ssh`` over the network (not using the FTDI
USB interface on the board), with the dev board and the host computer
connected to the same network. 

Assuming you are using the default ``petalinux`` user from the Xilinx
reference image , open a terminal on the host machine and ssh into the
development board with the boards IP address (found via ``ifconfig``):
```

**Comment by uLipe** [comment]
Does anyone know what is the CI error cause?

**Comment by microbuilder** [comment]
@henrikbrixandersen Mind taking another look? The current code should resolve your change request.

**Review by microbuilder — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]
Looks good, thanks!

**Review by carlocaione — APPROVED** [review]

**Comment by microbuilder** [comment]
Tested on HW with the Xilinx distro. Some additions to this platform are useful in the future, but current code looks good to me as a starting point.

**Comment by carlescufi** [comment]
CI failure has been fixed in main

**Pull request merged**
